docs/source/cpp/dataset.rst

```diff
@@ -0,0 +1,381 @@
+
+.. code-block:: text
+
+   dataset_name/
+     year=2007/
+       month=01/
+          data0.parquet
+          data1.parquet
+          ...
+       month=02/
+          data0.parquet
+          data1.parquet
+          ...
+       month=03/
+       ...
+     year=2008/
+       month=01/
+       ...
+     ...
+
+The above partitioning scheme is using "/key=value/" directory names, as found
+in Apache Hive.
+
+Let's create a small partitioned dataset. For this, we'll use Dataset's writing
+functionality.
+
+.. literalinclude:: ../../../cpp/examples/arrow/dataset_documentation_example.cc
+   :language: cpp
+   :lines: 106-149
+   :emphasize-lines: 25-42
+   :linenos:
+   :lineno-match:
+
+The above created a directory with two subdirectories ("part=a" and "part=b"),
+and the Parquet files written in those directories no longer include the "part"
+column.
+
+Reading this dataset, we now specify that the dataset uses a Hive-like
+partitioning scheme:
+
+.. literalinclude:: ../../../cpp/examples/arrow/dataset_documentation_example.cc
+   :language: cpp
+   :lines: 256-279
+   :emphasize-lines: 7,9-11
+   :linenos:
+   :lineno-match:
+
+Although the partition fields are not included in the actual Parquet files,
+they will be added back to the resulting table when scanning this dataset:
+
+.. code-block:: text
+
+   $ ./debug/dataset_documentation_example file:///tmp parquet_hive partitioned
+   Found fragment: /tmp/parquet_dataset/part=a/part0.parquet
+   Partition expression: (part == "a")
+   Found fragment: /tmp/parquet_dataset/part=b/part1.parquet
+   Partition expression: (part == "b")
+   Read 20 rows
+   a: int64
+     -- field metadata --
+     PARQUET:field_id: '1'
+   b: double
+     -- field metadata --
+     PARQUET:field_id: '2'
+   c: int64
+     -- field metadata --
+     PARQUET:field_id: '3'
+   part: string
+   ----
+   # snip...
+
+We can now filter on the partition keys, which avoids loading files
+altogether if they do not match the predicate:
+
+.. literalinclude:: ../../../cpp/examples/arrow/dataset_documentation_example.cc
+   :language: cpp
+   :lines: 281-301
+   :emphasize-lines: 15-18
+   :linenos:
+   :lineno-match:
+
+Different partitioning schemes
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The above example uses a hive-like directory scheme, such as "/year=2009/month=11/day=15".
+We specified this by passing the partition *factory*. In this case, the types of
+the partition keys are inferred from the file paths.
+
+It is also possible directly construct the partitioning and explicitly define
+the schema of the partition keys. For example:
+
+.. code-block:: cpp
+
+    auto part = std::make_shared<ds::HivePartitioning>(arrow::schema({
+        arrow::field("year", arrow::int16()),
+        arrow::field("month", arrow::int8()),
+        arrow::field("day", arrow::int32())
+    }));
+
+"Directory partitioning" is also supported, where the segments in the file path
```

Comment:
"Directory partitioning" is kind of a confusing name when "partitioning" means "using directories".  Maybe to help with that we could rephrase this to...

```
Arrow supports another form of partitioning, "Directory partitioning", where the segments...
```